src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```
@@ -605,6 +731,107 @@ public void withLock(boolean waitForRecovery, ChannelConsumer consumer) throws I
+                        if (!tempChannel.isOpen() && tempChannel.getConnection().isOpen()) {
+                            tempChannel = recreateChannel();
+                        }
+                    }
+                }
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        public <T> T retryingConsumeWithLock(ChannelMapper<T> mapper, ConnectionManagerConfiguration configuration) throws InterruptedException, IOException {
+            Iterator<RetryingDelay> iterator = null;
+            IOException exception;
+            Channel tempChannel = null;
+            boolean isChannelClosed = false;
+            while (true) {
+                lock.lock();
+                try {
+                    if (tempChannel == null) {
+                        tempChannel = getChannel();
+                    } else if (isChannelClosed) {
+                        tempChannel = recreateChannel();
+                    }
+
+                    var tag = mapper.map(tempChannel);
+                    isSubscribed = true;
+                    return tag;
+                } catch (IOException e) {
+                    var reason = tempChannel.getCloseReason();
+                    isChannelClosed = reason != null;
+                    if (isChannelClosed && reason.getMessage().contains("reply-text=RESOURCE_LOCKED")) {
+                        throw e;
+                    }
+                    exception = e;
+                } finally {
+                    lock.unlock();
+                }
+                iterator = handleAndSleep(configuration, iterator, "Retrying consume", exception);
+            }
+        }
+
+        public Optional<SubscriptionCallbacks> getCallbacksForRecovery(Channel channelToRecover) {
+            lock.lock();
+            SubscriptionCallbacks callbacks = null;
+            try {
+                if (channel != channelToRecover) {
+                    LOGGER.warn("Channel already recovered");
```

This case looks like we have bug in our code.
Maybe throw an exception in this case

---

Exception will not be handled because this code executed in executor service.

---

Entry point of our tasks should have try ... catch block to catch all Exceptions.